src/engine/prague.md

```diff
@@ -27,76 +22,21 @@ This specification is based on and extends [Engine API - Cancun](./cancun.md) sp
-
-### WithdrawalRequestV1
-This structure maps onto the withdrawal request from [EIP-7002](https://eips.ethereum.org/EIPS/eip-7002).
-The fields are encoded as follows:
-
-- `sourceAddress`: `DATA`, 20 Bytes
-- `validatorPubkey`: `DATA`, 48 Bytes
-- `amount`: `QUANTITY`, 64 Bits
-
-*Note:* The `amount` value is represented in Gwei.
-
-### ConsolidationRequestV1
-This structure maps onto the consolidation request from [EIP-7251](https://eips.ethereum.org/EIPS/eip-7251).
-The fields are encoded as follows:
-
-- `sourceAddress`: `DATA`, 20 Bytes
-- `sourcePubkey`: `DATA`, 48 Bytes
-- `targetPubkey`: `DATA`, 48 Bytes
-
-### ExecutionPayloadV4
-
-This structure has the syntax of [`ExecutionPayloadV3`](./cancun.md#executionpayloadv3) and appends the new fields: `depositRequests` and `withdrawalRequests`.
-
-- `parentHash`: `DATA`, 32 Bytes
-- `feeRecipient`:  `DATA`, 20 Bytes
-- `stateRoot`: `DATA`, 32 Bytes
-- `receiptsRoot`: `DATA`, 32 Bytes
-- `logsBloom`: `DATA`, 256 Bytes
-- `prevRandao`: `DATA`, 32 Bytes
-- `blockNumber`: `QUANTITY`, 64 Bits
-- `gasLimit`: `QUANTITY`, 64 Bits
-- `gasUsed`: `QUANTITY`, 64 Bits
-- `timestamp`: `QUANTITY`, 64 Bits
-- `extraData`: `DATA`, 0 to 32 Bytes
-- `baseFeePerGas`: `QUANTITY`, 256 Bits
-- `blockHash`: `DATA`, 32 Bytes
-- `transactions`: `Array of DATA` - Array of transaction objects, each object is a byte list (`DATA`) representing `TransactionType || TransactionPayload` or `LegacyTransaction` as defined in [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718)
-- `withdrawals`: `Array of WithdrawalV1` - Array of withdrawals, each object is an `OBJECT` containing the fields of a `WithdrawalV1` structure.
-- `blobGasUsed`: `QUANTITY`, 64 Bits
-- `excessBlobGas`: `QUANTITY`, 64 Bits
-- `depositRequests`: `Array of DepositRequestV1` - Array of deposits, each object is an `OBJECT` containing the fields of a `DepositRequestV1` structure.
-- `withdrawalRequests`: `Array of WithdrawalRequestV1` - Array of withdrawal requests, each object is an `OBJECT` containing the fields of a `WithdrawalRequestV1` structure.
-- `consolidationRequests`: `Array of ConsolidationRequestV1` - Array of consolidation requests, each object is an `OBJECT` containing the fields of a `ConsolidationRequestV1` structure.
-
 ## Methods
 
 ### engine_newPayloadV4
 
-The request of this method is updated with [`ExecutionPayloadV4`](#ExecutionPayloadV4).
+Method parameter list is extended with `executionRequestsHash`.
 
 #### Request
 
 * method: `engine_newPayloadV4`
 * params:
-  1. `executionPayload`: [`ExecutionPayloadV4`](#ExecutionPayloadV4).
+  1. `executionPayload`: [`ExecutionPayloadV3`](./cancun.md#executionpayloadv3).
   2. `expectedBlobVersionedHashes`: `Array of DATA`, 32 Bytes - Array of expected blob versioned hashes to validate.
   3. `parentBeaconBlockRoot`: `DATA`, 32 Bytes - Root of the parent beacon block.
+  4. `executionRequestsHash`: `DATA`, 32 Bytes - Hash of the execution layer triggered requests,
```

Comment:
nice! :+1:

Reply:
With the proposed design, the EL `executionRequestsHash` here is _NOT_ the same as `hash_tree_root(beacon_block.body.execution_requests)`, so the CL would have to do extra hashing effort to do a `newPayload`. See https://github.com/ethereum/EIPs/pull/8916#discussion_r1784309417


I think there are two ways that make sense:

- Either send the full requests here (same format as `getPayload`), and have the EL compute the EL specific `requestsHash`
- Or change the EL `requestsHash` to match `htr(ExecutionRequests)` (structure of request data is no longer opaque in the EL), so that the extra hashing effort is gone.

Reply:
+1 to Etan's comment, we definitely don't want the CL be computing this hash and we definitely don't want to be RLP encoding anything on the CL to compute this hash.

Reply:
also +1, left a similar comment here https://github.com/ethereum/consensus-specs/pull/3950#pullrequestreview-2338504099

I think the CL should just pass the full requests data and then the EL can do whatever it needs